### ISSUE TITLE
win10+A40 48G 通过pt微调

### ISSUE THREAD

**Issue by ptaam** [THUDM/ChatGLM3#305]
### System Info / 系統信息

win10系统，英伟达A40（48）显卡，cuda12.1，电脑内存32,G，i913代的cpu
chatglm3模型可以运行正常。
通过git bash的命令行启动微调代码中的finetune_pt.sh进行微调，模型已经下载到本地，训练数据在600条左右

### Who can help? / 谁可以帮助到您？

请教两个问题
1.目前A40微调起来占用内存大概在16G左右，显卡温度会很高在92-97之间拍换，这样的显卡温度是否正常，是否会影响微调效率
2.现在微调跑起来的速度很麻，大约140s/it，总共1000 大概需要38个小时，我的训练数据目前只有600条，请问这样的速度是否正常，内存和显存空间都很充足

### Information / 问题信息

- [X] The official example scripts / 官方的示例脚本
- [ ] My own modified scripts / 我自己修改的脚本和任务

### Reproduction / 复现过程

请教两个问题
1.目前A40微调起来占用内存大概在16G左右，显卡温度会很高在92-97之间拍换，这样的显卡温度是否正常，是否会影响微调效率
2.现在微调跑起来的速度很麻，大约140s/it，总共1000 大概需要38个小时，我的训练数据目前只有600条，请问这样的速度是否正常，内存和显存空间都很充足

### Expected behavior / 期待表现

请教两个问题
1.目前A40微调起来占用内存大概在16G左右，显卡温度会很高在92-97之间拍换，这样的显卡温度是否正常，是否会影响微调效率
2.现在微调跑起来的速度很麻，大约140s/it，总共1000 大概需要38个小时，我的训练数据目前只有600条，请问这样的速度是否正常，内存和显存空间都很充足

**Comment by zRzRzRzRzRzRzR** [comment]
配置肯定是没有问题，如果你是用了全参微调，那可能需要
如果你是p-tuning那肯定不对
至于显卡温度过高是你散热问题？

**Comment by ptaam** [comment]
/
/#! /usr/bin/env bash

set -ex

PRE_SEQ_LEN=128
LR=2e-2
NUM_GPUS=1
MAX_SOURCE_LEN=1024
MAX_TARGET_LEN=128
DEV_BATCH_SIZE=1
GRAD_ACCUMULARION_STEPS=32
MAX_STEP=1000
SAVE_INTERVAL=500

DATESTR=`date +%Y%m%d-%H%M%S`
RUN_NAME=advertise_gen_pt

BASE_MODEL_PATH=d:\\chatglm-model\\chatglm3-6b
DATASET_PATH=formatted_data\\advertise_gen.jsonl
OUTPUT_DIR=output/${RUN_NAME}-${DATESTR}-${PRE_SEQ_LEN}-${LR}

mkdir -p $OUTPUT_DIR

torchrun --standalone --nnodes=1 --nproc_per_node=$NUM_GPUS finetune.py \
    --train_format input-output \
    --train_file $DATASET_PATH \
    --preprocessing_num_workers 1 \
    --model_name_or_path $BASE_MODEL_PATH \
    --output_dir $OUTPUT_DIR \
    --max_source_length $MAX_SOURCE_LEN \
    --max_target_length $MAX_TARGET_LEN \
    --per_device_train_batch_size $DEV_BATCH_SIZE \
    --gradient_accumulation_steps $GRAD_ACCUMULARION_STEPS \
    --max_steps $MAX_STEP \
    --logging_steps 1 \
    --save_steps $SAVE_INTERVAL \
    --learning_rate $LR \
    --pre_seq_len $PRE_SEQ_LEN 2>&1 | tee ${OUTPUT_DIR}/train.log

**是通过的p-turning，脚本如上，advertise_gen.jsonl有600条数据**


**跑起来显示：**
{'loss': 0.01, 'learning_rate': 0.011080000000000001, 'epoch': 23.83}
{'loss': 0.0042, 'learning_rate': 0.01106, 'epoch': 23.88}/it]
{'loss': 0.0035, 'learning_rate': 0.011040000000000001, 'epoch': 23.93}
{'loss': 0.0037, 'learning_rate': 0.01102, 'epoch': 23.99}/it]
{'loss': 0.0038, 'learning_rate': 0.011000000000000001, 'epoch': 24.04}
 450/1000 [19:56:18<23:11:38, 151.81s/it](glm3)

**Comment by zRzRzRzRzRzRzR** [comment]
麻烦提供一下软件环境，包括cuda，Transformer等包环境和系统环境

**Comment by ptaam** [comment]
**系统**：w0专业版
**硬件**：cpu：i9-13900K   内存：32G  显卡：A40(48G)
**运行**：通过git bash命令行执行finetune_pt.sh 训练数据是输入/输出格式  跑起来内存大约占用12G 显存占用16G左右  资源充足 显卡使用功率在130W左右
**python版本**：3.9.11 （通过conda创建虚拟环境）
**cuda版本**：12.1
**环境**：
absl-py                   2.0.0
accelerate                0.24.1
aiofiles                  23.2.1
aiohttp                   3.8.6
aiosignal                 1.3.1
altair                    5.1.2
anyio                     4.0.0
astunparse                1.6.3
async-timeout             4.0.3
attrs                     23.1.0
blinker                   1.7.0
Brotli                    1.0.9
cachetools                5.3.2
certifi                   2023.7.22
cffi                      1.15.1
charset-normalizer        2.0.4
click                     8.1.7
colorama                  0.4.6
contourpy                 1.1.1
cpm-kernels               1.0.11
cryptography              41.0.3
cycler                    0.12.1
deepspeed                 0.3.16
exceptiongroup            1.1.3
fastapi                   0.95.1
ffmpy                     0.3.1
filelock                  3.9.0
fonttools                 4.43.1
frozenlist                1.4.0
fsspec                    2023.10.0
gitdb                     4.0.11
GitPython                 3.1.40
google-auth               2.23.4
google-auth-oauthlib      1.1.0
gradio                    3.39.0
gradio_client             0.7.0
grpcio                    1.59.2
h11                       0.14.0
httpcore                  0.18.0
httpx                     0.25.0
huggingface-hub           0.18.0
idna                      3.4
importlib-metadata        6.8.0
importlib-resources       6.1.0
Jinja2                    3.1.2
jsonschema                4.19.2
jsonschema-specifications 2023.7.1
kiwisolver                1.4.5
latex2mathml              3.76.0
linkify-it-py             2.0.2
Markdown                  3.5.1
markdown-it-py            2.2.0
MarkupSafe                2.1.1
matplotlib                3.8.1
mdit-py-plugins           0.3.3
mdtex2html                1.2.0
mdurl                     0.1.2
mkl-fft                   1.3.8
mkl-random                1.2.4
mkl-service               2.4.0
mpmath                    1.3.0
multidict                 6.0.4
networkx                  3.1
ninja                     1.11.1.1
numpy                     1.26.0
oauthlib                  3.2.2
orjson                    3.9.10
packaging                 23.2
pandas                    2.1.2
Pillow                    10.0.1
pip                       23.3
protobuf                  4.23.4
psutil                    5.9.6
pyarrow                   14.0.0
pyasn1                    0.5.0
pyasn1-modules            0.3.0
pycparser                 2.21
pydantic                  1.10.13
pydeck                    0.8.1b0
pydub                     0.25.1
Pygments                  2.16.1
pyOpenSSL                 23.2.0
pyparsing                 3.1.1
PySocks                   1.7.1
python-dateutil           2.8.2
python-multipart          0.0.6
pytz                      2023.3.post1
PyYAML                    6.0.1
referencing               0.30.2
regex                     2023.10.3
requests                  2.31.0
requests-oauthlib         1.3.1
rich                      13.6.0
rpds-py                   0.10.6
rsa                       4.9
safetensors               0.4.0
semantic-version          2.10.0
sentencepiece             0.1.99
setuptools                68.0.0
six                       1.16.0
smmap                     5.0.1
sniffio                   1.3.0
sse-starlette             1.6.5
starlette                 0.26.1
streamlit                 1.28.0
sympy                     1.11.1
tenacity                  8.2.3
tensorboard               2.15.1
tensorboard-data-server   0.7.2
tensorboardX              1.8
tokenizers                0.13.3
toml                      0.10.2
toolz                     0.12.0
torch                     2.1.0
torchaudio                2.1.0
torchvision               0.16.0
tornado                   6.3.3
tqdm                      4.66.1
transformers              4.30.2
typing_extensions         4.4.0
tzdata                    2023.3
tzlocal                   5.2
uc-micro-py               1.0.2
urllib3                   1.26.18
uvicorn                   0.23.2
validators                0.22.0
watchdog                  3.0.0
websockets                11.0.3
Werkzeug                  3.0.1
wheel                     0.41.2
win-inet-pton             1.1.0
yarl                      1.9.2
zipp                      3.17.0

**训练数据格式如下**：
{"prompt": "提取*能辉科技发力重卡换电市场,自主研发换电机器人面市。*中心词", "response": "重卡、换电"}
{"prompt": "提取*安徽省出台农业机械化“十四五”发展规划。*中心词", "response": "农业机械"}
{"prompt": "提取*反垄断法修正草案、反电信网络诈骗法草案等8部法律草案将于近期审议。*中心词", "response": "反电信网络诈骗"}
{"prompt": "提取*住房公积金数字人民币应用场景在广州正式落地。*中心词", "response": "数字人民币"}
{"prompt": "提取*荣联科技成功中标国家开发银行网络安全项目。*中心词", "response": "开发银行网络安全、荣联科技"}
{"prompt": "提取*深圳加快培育鸿蒙欧拉生态,最高资助1000万元。*中心词", "response": "鸿蒙欧拉、鸿蒙生态"}

**Comment by ptaam** [comment]
win10专业版

**Comment by zRzRzRzRzRzRzR** [comment]
mark，我们验证一下，我们没有A40的显卡，我们会用其他显卡验证并告诉你结果

**Comment by ptaam** [comment]
好的

**Comment by ptaam** [comment]
我这斌同样的配置同样的600条数据集尝试了对chatglm2进行微调，发现速度在15s/it，共3000个预计12个小时左右，显存占用也是在16G左右，这个时间是否属于该数据集的正常范围？
train.sh文件如下：
PRE_SEQ_LEN=128
LR=2e-2
NUM_GPUS=1

torchrun --standalone --nnodes=1 --nproc-per-node=$NUM_GPUS main.py \
    --do_train \
    --train_file formatted_data\\advertise_gen.json \
    --validation_file formatted_data\\dev.json \
    --preprocessing_num_workers 10 \
    --prompt_column prompt \
    --response_column response \
    --overwrite_cache \
    --model_name_or_path D:\\chatglm-model\\chatglm2-6b \
    --output_dir output\\adgen-chatglm2-6b-pt-$PRE_SEQ_LEN-$LR \
    --overwrite_output_dir \
    --max_source_length 64 \
    --max_target_length 128 \
    --per_device_train_batch_size 1 \
    --per_device_eval_batch_size 1 \
    --gradient_accumulation_steps 16 \
    --predict_with_generate \
    --max_steps 3000 \
    --logging_steps 10 \
    --save_steps 1000 \
    --learning_rate $LR \
    --pre_seq_len $PRE_SEQ_LEN

**Comment by Btlmd** [comment]
> 1.目前A40微调起来占用内存大概在16G左右，显卡温度会很高在92-97之间拍换，这样的显卡温度是否正常，是否会影响微调效率
2.现在微调跑起来的速度很麻，大约140s/it，总共1000 大概需要38个小时，我的训练数据目前只有600条，请问这样的速度是否正常，内存和显存空间都很充足

1. 温度较高。这可能是散热的问题。
2. 140s/it 不太正常。可以使用  `nvidia-smi` 检查一下显卡的功率和利用率。可能是温度太高导致的效率下降？
